lua/ttt2/libraries/database.lua

```diff
@@ -0,0 +1,1419 @@
+-- @param string accessName the chosen accessName registered for a given database. HAS NOT TO BE the real database-name!
+-- @param[opt] string itemName the name of the item in the database. Leave `nil` if you want to remove callbacks for every item with given identifier
+-- @param[opt] string key the name of the key in the database. Leave `nil` if you want to remove callbacks for every key with given identifier
+-- @param[opt] string identifier a chosen identifier if you want to remove the callback
+-- @realm shared
+function database.RemoveChangeCallback(accessName, itemName, key, identifier)
+	callbacks = callbackIdentifiers[identifier]
+
+	if not accessName or not istable(callbacks) then return end
+
+	local cache = callbackCache[accessName]
+
+	if not istable(cache) then return end
+
+	-- If no itemName is given, remove callbacks of all items
+	local skipWrongItemName = true
+	if not isstring(itemName) then
+		skipWrongItemName = false
+	end
+
+	-- If no key is given, remove callbacks of all keys of the item	
+	local skipWrongKey = true
+	if not isstring(key) then
+		skipWrongKey = false
+	end
+
+	for i = #callbacks, 1, -1  do
+		callback = callbacks[i]
+
+		-- AccesName has to be the same, because registrating callbacks for all sql tables is not allowed
+		if callback.accessName ~= accessName then continue end
+
+		-- If neither itemName nor the key fit, skip that callback unless you want to remove all
+		local cItemName = callback.itemName
+		if cItemName ~= itemName and skipWrongItemName then continue end
+
+		local cKey = callback.key
+		if cKey ~= key and skipWrongKey then continue end
+
+		table.remove(callbacks, i)
+
+		if not istable(cache[cItemName]) or not istable(cache[cItemName][cKey]) then continue end
+		cache[cItemName][cKey][identifier] = nil
+
+		-- Delete empty table entries in callbackCache
+		if not table.IsEmpty(cache[cItemName][cKey]) then continue end
+		cache[cItemName][cKey] = nil
+
+		if not table.IsEmpty(cache[cItemName]) then continue end
+		cache[cItemName] = nil
+
+		if not table.IsEmpty(cache) then continue end
+		callbackCache[accessName] = nil
+	end
+
+	-- If no callbacks for that identifier are left, delete it
+	if #callbacks < 1 then
+		callbackIdentifiers[identifier] = nil
+	end
+end
+
+---
+-- Converts the given value of a database with its key
+-- @warning this function should only be internally used and shall be removed and replaced with orm conversion. 
+-- `sql.GetParsedData` is not compatible with orm as sql.SQLStr is used (e.g. converts `false` => "false"
+-- while the other uses "0" and "1" for booleans
+-- @param string value the value to convert with a key
+-- @param string accessName the chosen accessName registered for a given database.
+-- @param string key the name of the key in the database
+-- @return any value after conversion, `nil` if not convertible or "nil" or "NULL"
+-- @realm shared
+-- @internal
+function database.ConvertValueWithKey(value, accessName, key)
```

Comment:
should this internal function be visble to the outside? or is it enough to have a local function do this?

Reply:
I guess not. I will re-check why I did that,

Reply:
I guess I thought I would use it outside, but I am not. Probably still an early design idea, I threw away, I localized it 😄